mlos_core/mlos_core/optimizers/optimizer.py

```diff
@@ -172,41 +178,43 @@ def register_pending(self, configurations: pd.DataFrame,
         """
         pass    # pylint: disable=unnecessary-pass # pragma: no cover
 
-    def get_observations(self) -> pd.DataFrame:
-        """Returns the observations as a dataframe.
+    def get_observations(self) -> Tuple[pd.DataFrame, pd.DataFrame, Optional[pd.DataFrame]]:
+        """
+        Returns the observations as a triplet of DataFrames (config, score, context).
 
         Returns
         -------
-        observations : pd.DataFrame
-            Dataframe of observations. The columns are parameter names and "score" for the score, each row is an observation.
+        observations : Tuple[pd.DataFrame, pd.DataFrame, Optional[pd.DataFrame]]
+            A triplet of (config, score, context) DataFrames of observations.
         """
         if len(self._observations) == 0:
             raise ValueError("No observations registered yet.")
-        configs = pd.concat([config for config, _, _ in self._observations])
-        scores = pd.concat([score for _, score, _ in self._observations])
-        try:
-            contexts = pd.concat([context for _, _, context in self._observations if context is not None])
-        except ValueError:
-            contexts = None
-        configs["score"] = scores
-        if contexts is not None:
-            # configs = pd.concat([configs, contexts], axis=1)
-            # Not reachable for now
-            raise NotImplementedError()
-        return configs
-
-    def get_best_observation(self) -> pd.DataFrame:
-        """Returns the best observation so far as a dataframe.
+        configs = pd.concat([config for config, _, _ in self._observations]).reset_index(drop=True)
+        scores = pd.concat([score for _, score, _ in self._observations]).reset_index(drop=True)
+        contexts = pd.concat([pd.DataFrame() if context is None else context
+                              for _, _, context in self._observations]).reset_index(drop=True)
+        return (configs, scores, contexts if len(contexts.columns) > 0 else None)
+
+    def get_best_observations(self, n_max: int = 1) -> Tuple[pd.DataFrame, pd.DataFrame, Optional[pd.DataFrame]]:
+        """
+        Get the N best observations so far as a triplet of DataFrames (config, score, context).
```

Comment:
Add a note on how these are ordered?

For instance, this may produce the pareto frontier, or more than that if N is large, correct?

Reply:
good point!